File: kattis8/floodit.py
from queue import *
udir = [0,0,-1,1]
vdir = [1,-1,0,0]
n = 0
def bfs(arr, num, sarr, coloring):
	q = Queue()
	vis = [[False for _ in range(n)] for _ in range(n)]
	for x in sarr:
		q.put(x)
		vis[x[0]][x[1]] = True
	co = 0
	while not q.empty():
		co += 1
		r, c = q.get()
		for i in range(4):
			nr = r + udir[i]
			nc = c + vdir[i]
			if nr in range(n) and nc in range(n) and not vis[nr][nc] and arr[nr][nc] == num:
				vis[nr][nc] = True
				q.put((nr,nc))
				if coloring:
					sarr.append((nr,nc))
	return co

tc = int(input())
for _ in range(tc):
	n = int(input())
	arr = []
	for _ in range(n):
		arr.append(list(map(int,list(input()))))
	marr = [0 for _ in range(6)]
	sarr = [(0,0)]
	bfs(arr, arr[0][0], sarr, True)
	while len(sarr) != n*n:
		curmx = (0, 0)
		for i in range(1,7):
			curmx = max(curmx, (bfs(arr, i, sarr, False),-i))
		bfs(arr, -curmx[1], sarr, True)
		marr[-curmx[1]-1] +=1
	print(sum(marr))
	print(' '.join(map(str, marr)))

File: codeforces/round 719/floodit.cpp
#define REP(i, n) for(int i = 0; i < n; i ++)
#define REPL(i,m, n) for(int i = m; i < n; i ++)
#define FOREACH(it, l) for (auto it = l.begin(); it != l.end(); it++)
#define SORT(arr) sort(arr.begin(), arr.end())
#define LSOne(S) ((S)&-(S))
#define M_PI 3.1415926535897932384
#define INF 999999999
#include <bits/stdc++.h>
#include <ext/pb_ds/assoc_container.hpp>
#include <ext/pb_ds/tree_policy.hpp>
using namespace __gnu_pbds;
using namespace std;
typedef vector<int> vi;
typedef vector<vi> vvi;
typedef pair<int, int> ii;
typedef vector<ii> vii;
typedef long long ll;
typedef vector<ll> vll;
typedef vector<vll> vvll;
typedef double ld;
int n;
vvi arr;
vi udir = {0,0,-1,1};
vi vdir = {1,-1,0,0};
int bfs(int num, vii &sarr, bool coloring){
	queue<ii> q;
	bool vis[n][n];
	memset(vis, 0, sizeof(vis));
	for (auto [r,c] : sarr) {
		q.push({r,c});
		vis[r][c] = 1;
	}
	int co = 0;
	while (!q.empty()) {
		co += 1;
		auto [r, c] = q.front();q.pop();
		REP(i, 4) {
			int nr = r + udir[i];
			int nc = c + vdir[i];
			if (nr>=0&&nr<n&&nc>=0&&nc<n && (!vis[nr][nc]) && arr[nr][nc] == num) {
				vis[nr][nc] =1;
				q.push({nr,nc});
				if (coloring) sarr.push_back({nr,nc});
			}
		}
	}
	return co;
}
void solve() {
	cin>>n;
	arr.assign(n, vi(n));
	REP(i, n) {
		string s;cin>>s;
		REP(j, n) {
			arr[i][j] = s[j]-'0';
		}
	}
	vi marr(6,0);
	int tot =0;
	vii sarr = {{0,0}};
	bfs(arr[0][0], sarr, true);
	while (sarr.size() != n*n){
		tot++;
		ii curmx = {0, 0};
		REPL(i, 1,7) {
			curmx = max(curmx, {bfs(i, sarr, false),-i});
		}
		bfs(-curmx.second, sarr, true);
		marr[-curmx.second-1] +=1;
	}
	cout<<tot<<"\n";
	for (auto i: marr) {
		cout<<i<<" ";
	}cout<<"\n";
}

int main()
{
	ios_base::sync_with_stdio(false);
	cin.tie(NULL); cout.tie(NULL);
	int tc;
	cin>>tc;
	REP(i, tc) {
		solve();
	}

}